### PR TITLE
Prefer length and percentage for word spacing

### PR DESCRIPTION
**Pull request by ghost**
The goal is to make use of `LengthOrPercentage` for word spacing in `ShapingOptions`, but since it makes use of `f32` which doesn't implement `Hash`, we're going for `NotNan<f32>` from [ordered-float](https://github.com/reem/rust-ordered-float/), which supports hashing. Instead of implementing `Hash` for `LengthOrPercentage` and thereby the inner types like `CSSFloat`, `CalcLengthOrPercentage`, etc., we convert it to `(Au, NotNan<f32>)`.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [ ] There are tests for these changes

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12783)

### CODE REVIEW

File: components/gfx/font.rs [hunk 1]
```diff
@@ -225,7 +226,8 @@ impl Font {
 
             let mut advance = Au::from_f64_px(self.glyph_h_advance(glyph_id));
             if character == ' ' {
-                advance += options.word_spacing;
+                let (length, percent) = options.word_spacing;
+                advance = (advance + length) + Au((advance.0 as f32 * percent.into_inner()) as i32);
```
Comment (Manishearth) [inline components/gfx/font.rs]: Link to https://www.w3.org/TR/css-text-3/#word-spacing

File: components/gfx/text/shaping/harfbuzz.rs [hunk 2]
```diff
@@ -403,7 +403,8 @@ impl Shaper {
         // applied. The effect of the property on other word-separator characters is undefined."
         // We elect to only space the two required code points.
         if character == ' ' || character == '\u{a0}' {
-            advance = advance + options.word_spacing
+            let (length, percent) = options.word_spacing;
+            advance = (advance + length) + Au((advance.0 as f32 * percent.into_inner()) as i32);
```
Comment (Manishearth) [inline components/gfx/text/shaping/harfbuzz.rs]: Ditto about the spec link

File: components/style/properties/longhand/inherited_text.mako.rs [hunk 3]
```diff
@@ -285,23 +285,23 @@
     #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
     pub enum SpecifiedValue {
         Normal,
-        Specified(specified::Length),  // FIXME(SimonSapin) support percentages
+        LengthOrPercentage(specified::LengthOrPercentage),
```
Comment (Manishearth) [inline components/style/properties/longhand/inherited_text.mako.rs]: Don't rename it from Specified, let it be `Specified(specified::Lop)`

**Comment by Manishearth** [comment]
LGTM modulo nits.

r=me post @SimonSapin's sign-off

**Comment by Manishearth** [comment]
Post IRC discussion we know that your method of calculation is correct.

r=me

needs lockfile updates:

```patch
diff --git a/ports/geckolib/Cargo.lock b/ports/geckolib/Cargo.lock
index 405fadd..b60e592 100644
--- a/ports/geckolib/Cargo.lock
+++ b/ports/geckolib/Cargo.lock
@@ -215,6 +215,33 @@ dependencies = [
 ]
 
 [[package]]
+name = "num"
+version = "0.1.32"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "num-integer 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+ "num-iter 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+ "num-traits 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "num-integer"
+version = "0.1.32"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "num-traits 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "num-iter"
+version = "0.1.32"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "num-integer 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+ "num-traits 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
 name = "num-traits"
 version = "0.1.32"
 source = "registry+https://github.com/rust-lang/crates.io-index"
@@ -228,6 +255,15 @@ dependencies = [
 ]
 
 [[package]]
+name = "ordered-float"
+version = "0.2.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "num 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+ "unreachable 0.0.2 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
 name = "quickersort"
 version = "2.0.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
@@ -325,6 +361,7 @@ dependencies = [
  "log 0.3.6 (registry+https://github.com/rust-lang/crates.io-index)",
  "matches 0.1.2 (registry+https://github.com/rust-lang/crates.io-index)",
  "num-traits 0.1.32 (registry+https://github.com/rust-lang/crates.io-index)",
+ "ordered-float 0.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
  "rand 0.3.14 (registry+https://github.com/rust-lang/crates.io-index)",
  "rustc-serialize 0.3.19 (registry+https://github.com/rust-lang/crates.io-index)",
  "selectors 0.7.0 (registry+https://github.com/rust-lang/crates.io-index)",
@@ -388,6 +425,14 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 
 [[package]]
 name = "unreachable"
+version = "0.0.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "void 0.0.5 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "unreachable"
 version = "0.1.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 dependencies = [
@@ -427,6 +472,11 @@ dependencies = [
 
 [[package]]
 name = "void"
+version = "0.0.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+
+[[package]]
+name = "void"
 version = "1.0.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 ```

**Comment by Wafflespeanut** [comment]
lock file updated :)
@bors-servo r=Manishearth

Thanks! :smile:

**Comment by Wafflespeanut** [comment]
forgot `ports/cef` - sorry

**Pull request merged**